File: src/flipbook_maker/flipbook/layout.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from PIL import Image, ImageColor, ImageDraw, ImageFont

from flipbook_maker.core.paper import A4_MM
from flipbook_maker.core.units import mm_to_px

FitMode = Literal["contain", "cover", "stretch"]


@dataclass(frozen=True)
class LayoutConfig:
    cols: int = 2
    rows: int = 8
    dpi: int = 300
    margin_mm: float = 5.0
    background: str | Path = "white"
    page_size_mm: tuple[float, float] = A4_MM
    cut_marks: bool = False
    cell_outline: bool = False
    fit: FitMode = "contain"
    frame_numbers: bool = False
    frame_number_color: str = "black"
    frame_number_offset_mm: float = 2.0
    bind_strip_mm: float = 0.0
    bind_strip_color: str | None = None

    def page_px(self) -> tuple[int, int]:
        return mm_to_px(self.page_size_mm[0], self.dpi), mm_to_px(self.page_size_mm[1], self.dpi)

    def margin_px(self) -> int:
        return mm_to_px(self.margin_mm, self.dpi)

    def cell_px(self) -> tuple[int, int]:
        w, h = self.page_px()
        m = self.margin_px()
        return (w - 2 * m) // self.cols, (h - 2 * m) // self.rows

    def bind_strip_px(self) -> int:
        return mm_to_px(self.bind_strip_mm, self.dpi)


def _load_background(spec: str | Path, size: tuple[int, int]) -> Image.Image:
    spec_path = Path(spec) if not isinstance(spec, Path) else spec
    if spec_path.exists() and spec_path.is_file():
        bg = Image.open(spec_path).convert("RGB")
        return bg.resize(size, Image.LANCZOS)
    try:
        rgb = ImageColor.getrgb(str(spec))
    except ValueError as e:
        msg = f"background must be a color name/hex or an image path, got: {spec!r}"
        raise ValueError(msg) from e
    return Image.new("RGB", size, rgb)


def _place_in_cell(
    cell: Image.Image, frame: Image.Image, fit: FitMode = "contain", bind_strip_px: int = 0
) -> None:
    cw, ch = cell.size
    fw, fh = frame.size
    usable_w = max(1, cw - bind_strip_px)

    if fit == "contain":
        scale = min(usable_w / fw, ch / fh)
        new = (max(1, round(fw * scale)), max(1, round(fh * scale)))
        resized = frame.resize(new, Image.LANCZOS)
        x = cw - new[0]
        y = (ch - new[1]) // 2
        if resized.mode == "RGBA":
            cell.paste(resized, (x, y), resized)
        else:
            cell.paste(resized, (x, y))

    elif fit == "cover":
        scale = max(usable_w / fw, ch / fh)
        crop_w = usable_w / scale
        crop_h = ch / scale
        left = round(fw - crop_w)
        top = round((fh - crop_h) / 2)
        cropped = frame.crop((left, top, fw, fh - top))
        resized = cropped.resize((usable_w, ch), Image.LANCZOS)
        if resized.mode == "RGBA":
            cell.paste(resized, (bind_strip_px, 0), resized)
        else:
            cell.paste(resized, (bind_strip_px, 0))

    elif fit == "stretch":
        resized = frame.resize((usable_w, ch), Image.LANCZOS)
        if resized.mode == "RGBA":
            cell.paste(resized, (bind_strip_px, 0), resized)
        else:
            cell.paste(resized, (bind_strip_px, 0))


def _draw_frame_number(
    page: Image.Image,
    frame_idx: int,
    cell_x: int,
    cell_y: int,
    cell_h: int,
    config: LayoutConfig,
) -> None:
    font = ImageFont.load_default()
    label = str(frame_idx)
    offset_px = mm_to_px(config.frame_number_offset_mm, config.dpi)
    bind_px = mm_to_px(config.bind_strip_mm, config.dpi)
    # When there's an explicit strip, keep numbers visible inside it.
    # Use the larger of the default offset and half the strip width so the
    # number sits comfortably in wider strips without being at the wall.
    effective_offset_px = max(offset_px, bind_px // 2) if bind_px > 0 else offset_px
    draw = ImageDraw.Draw(page)
    bbox = draw.textbbox((0, 0), label, font=font)
    text_h = bbox[3] - bbox[1]
    text_x = cell_x + effective_offset_px
    text_y = cell_y + cell_h // 2 - text_h // 2
    draw.text(
        (text_x, text_y),
        label,
        font=font,
        fill=ImageColor.getrgb(config.frame_number_color),
    )


def render_sheets(frames: list[Path], config: LayoutConfig) -> list[Image.Image]:
    page_size = config.page_px()
    cell_w, cell_h = config.cell_px()
    margin = config.margin_px()
    per_page = config.cols * config.rows
    bind_strip_px = config.bind_strip_px()

    pages: list[Image.Image] = []
    for start in range(0, len(frames), per_page):
        batch = frames[start : start + per_page]
        page = _load_background(config.background, page_size)
        for i, frame_path in enumerate(batch):
            row, col = divmod(i, config.cols)
            cell = Image.new("RGBA", (cell_w, cell_h), (0, 0, 0, 0))
            if bind_strip_px > 0 and config.bind_strip_color is not None:
                strip_rgb = ImageColor.getrgb(config.bind_strip_color)
                strip_region = Image.new("RGBA", (bind_strip_px, cell_h), strip_rgb + (255,))
                cell.paste(strip_region, (0, 0))
            with Image.open(frame_path) as raw:
                frame = raw.convert("RGBA")
                _place_in_cell(cell, frame, config.fit, bind_strip_px)
            x = margin + col * cell_w
            y = margin + row * cell_h
            page.paste(cell, (x, y), cell)
            if config.frame_numbers:
                _draw_frame_number(page, start + i + 1, x, y, cell_h, config)

        if config.cut_marks or config.cell_outline:
            page_w, page_h = page_size
            draw = ImageDraw.Draw(page)
            tick_len = mm_to_px(3.0, config.dpi)
            mark_w = max(1, mm_to_px(0.25, config.dpi))
            outline_w = max(1, mm_to_px(0.2, config.dpi))

            for row in range(config.rows):
                for col in range(config.cols):
                    cx = margin + col * cell_w
                    cy = margin + row * cell_h

                    if config.cell_outline:
                        draw.rectangle(
                            [(cx, cy), (cx + cell_w, cy + cell_h)],
                            outline=(0, 0, 0),
                            width=outline_w,
                        )

                    if config.cut_marks:
                        corners = [
                            (cx, cy, -1, -1),
                            (cx + cell_w, cy, 1, -1),
                            (cx, cy + cell_h, -1, 1),
                            (cx + cell_w, cy + cell_h, 1, 1),
                        ]
                        for x, y, hd, vd in corners:
                            hx = max(0, min(page_w - 1, x + hd * tick_len))
                            draw.line([(hx, y), (x, y)], fill=(0, 0, 0), width=mark_w)
                            vy = max(0, min(page_h - 1, y + vd * tick_len))
                            draw.line([(x, vy), (x, y)], fill=(0, 0, 0), width=mark_w)

        pages.append(page)
    return pages

File: web/src/core/layout.ts
import { mmToPx } from "./units";

export type FitMode = "contain" | "cover" | "stretch";

export interface LayoutConfig {
  cols: number;
  rows: number;
  dpi: number;
  marginMm: number;
  background: string;
  pageSizeMm: [number, number];
  cutMarks: boolean;
  cellOutline: boolean;
  fit: FitMode;
  frameNumbers: boolean;
  frameNumberColor: string;
  frameNumberOffsetMm: number;
  bindStripMm: number;
  bindStripColor: string | null;
}

export const DEFAULT_CONFIG: LayoutConfig = {
  cols: 2,
  rows: 8,
  dpi: 150,
  marginMm: 5,
  background: "white",
  pageSizeMm: [210, 297],
  cutMarks: false,
  cellOutline: false,
  fit: "contain",
  frameNumbers: false,
  frameNumberColor: "black",
  frameNumberOffsetMm: 2,
  bindStripMm: 0,
  bindStripColor: null,
};

export interface LayoutDimensions {
  pagePx: [number, number];
  marginPx: number;
  cellPx: [number, number];
  bindStripPx: number;
  usableW: number;
}

export function computeDimensions(config: LayoutConfig): LayoutDimensions {
  const pagePx: [number, number] = [
    mmToPx(config.pageSizeMm[0], config.dpi),
    mmToPx(config.pageSizeMm[1], config.dpi),
  ];
  const marginPx = mmToPx(config.marginMm, config.dpi);
  const cellPx: [number, number] = [
    Math.floor((pagePx[0] - 2 * marginPx) / config.cols),
    Math.floor((pagePx[1] - 2 * marginPx) / config.rows),
  ];
  const bindStripPx = mmToPx(config.bindStripMm, config.dpi);
  const usableW = cellPx[0] - bindStripPx;
  return { pagePx, marginPx, cellPx, bindStripPx, usableW };
}

function makeCanvas(w: number, h: number): HTMLCanvasElement {
  const c = document.createElement("canvas");
  c.width = w;
  c.height = h;
  return c;
}

export async function fillBackground(
  ctx: CanvasRenderingContext2D,
  bg: string,
  w: number,
  h: number,
): Promise<void> {
  if (bg.startsWith("data:") || bg.startsWith("blob:") || bg.startsWith("http")) {
    await new Promise<void>((resolve, reject) => {
      const img = new Image();
      img.onload = () => {
        ctx.drawImage(img, 0, 0, w, h);
        resolve();
      };
      img.onerror = reject;
      img.src = bg;
    });
  } else {
    ctx.fillStyle = bg;
    ctx.fillRect(0, 0, w, h);
  }
}

export async function renderCell(
  frame: ImageBitmap,
  frameIndex: number,
  config: LayoutConfig,
  dims: LayoutDimensions,
): Promise<HTMLCanvasElement> {
  const [cw, ch] = dims.cellPx;
  const canvas = makeCanvas(cw, ch);
  const ctx = canvas.getContext("2d")!;

  await fillBackground(ctx, config.background, cw, ch);

  if (dims.bindStripPx > 0 && config.bindStripColor) {
    ctx.fillStyle = config.bindStripColor;
    ctx.fillRect(0, 0, dims.bindStripPx, ch);
  }

  const fw = frame.width;
  const fh = frame.height;
  const uw = dims.usableW;

  if (config.fit === "contain") {
    const scale = Math.min(uw / fw, ch / fh);
    const pw = Math.round(fw * scale);
    const ph = Math.round(fh * scale);
    const px = cw - pw;
    const py = Math.round((ch - ph) / 2);
    ctx.drawImage(frame, px, py, pw, ph);
  } else if (config.fit === "cover") {
    const scale = Math.max(uw / fw, ch / fh);
    const cropW = Math.round(uw / scale);
    const cropH = Math.round(ch / scale);
    const cropX = Math.round((fw - cropW) / 2);
    const cropY = Math.round((fh - cropH) / 2);
    ctx.drawImage(frame, cropX, cropY, cropW, cropH, dims.bindStripPx, 0, uw, ch);
  } else {
    // stretch
    ctx.drawImage(frame, dims.bindStripPx, 0, uw, ch);
  }

  if (config.frameNumbers) {
    const offsetPx = mmToPx(config.frameNumberOffsetMm, config.dpi);
    const effectiveOffset =
      dims.bindStripPx > 0
        ? Math.max(offsetPx, Math.floor(dims.bindStripPx / 2))
        : offsetPx;
    const fontSize = Math.max(8, Math.round(ch * 0.06));
    ctx.font = `${fontSize}px sans-serif`;
    ctx.fillStyle = config.frameNumberColor;
    const text = String(frameIndex + 1);
    const metrics = ctx.measureText(text);
    const textH =
      metrics.actualBoundingBoxAscent + metrics.actualBoundingBoxDescent;
    ctx.fillText(text, effectiveOffset, (ch - textH) / 2 + metrics.actualBoundingBoxAscent);
  }

  return canvas;
}

export function drawCutMarks(
  ctx: CanvasRenderingContext2D,
  config: LayoutConfig,
  dims: LayoutDimensions,
): void {
  const tickPx = mmToPx(3, config.dpi);
  ctx.strokeStyle = "black";
  ctx.lineWidth = 1;

  for (let row = 0; row <= config.rows; row++) {
    for (let col = 0; col <= config.cols; col++) {
      const x = dims.marginPx + col * dims.cellPx[0];
      const y = dims.marginPx + row * dims.cellPx[1];

      ctx.beginPath();
      ctx.moveTo(x - tickPx, y);
      ctx.lineTo(x + tickPx, y);
      ctx.stroke();

      ctx.beginPath();
      ctx.moveTo(x, y - tickPx);
      ctx.lineTo(x, y + tickPx);
      ctx.stroke();
    }
  }
}

export function drawCellOutlines(
  ctx: CanvasRenderingContext2D,
  config: LayoutConfig,
  dims: LayoutDimensions,
): void {
  ctx.strokeStyle = "black";
  ctx.lineWidth = 1;
  for (let row = 0; row < config.rows; row++) {
    for (let col = 0; col < config.cols; col++) {
      const x = dims.marginPx + col * dims.cellPx[0];
      const y = dims.marginPx + row * dims.cellPx[1];
      ctx.strokeRect(x, y, dims.cellPx[0], dims.cellPx[1]);
    }
  }
}

export async function renderSheets(
  frames: ImageBitmap[],
  config: LayoutConfig,
  onProgress?: (current: number, total: number) => void,
): Promise<HTMLCanvasElement[]> {
  const dims = computeDimensions(config);
  const cellsPerPage = config.cols * config.rows;
  const pageCount = Math.max(1, Math.ceil(frames.length / cellsPerPage));
  const pages: HTMLCanvasElement[] = [];

  for (let p = 0; p < pageCount; p++) {
    const canvas = makeCanvas(dims.pagePx[0], dims.pagePx[1]);
    const ctx = canvas.getContext("2d")!;

    await fillBackground(ctx, config.background, dims.pagePx[0], dims.pagePx[1]);

    const startIdx = p * cellsPerPage;
    const endIdx = Math.min(startIdx + cellsPerPage, frames.length);

    for (let i = startIdx; i < endIdx; i++) {
      const cellIndex = i - startIdx;
      const col = cellIndex % config.cols;
      const row = Math.floor(cellIndex / config.cols);
      const x = dims.marginPx + col * dims.cellPx[0];
      const y = dims.marginPx + row * dims.cellPx[1];

      const cell = await renderCell(frames[i], i, config, dims);
      ctx.drawImage(cell, x, y);

      if (onProgress) onProgress(i + 1, frames.length);
    }

    if (config.cutMarks) drawCutMarks(ctx, config, dims);
    if (config.cellOutline) drawCellOutlines(ctx, config, dims);

    pages.push(canvas);
  }

  return pages;
}
